File: lena-config/src/main/java/org/riversoforion/lena/config/PrioritizedConfigurationSource.java
/*
 * Copyright (c) 2024-2025. Eric McIntyre / Rivers of Orion
 */
package org.riversoforion.lena.config;

import java.util.List;
import java.util.SortedMap;
import java.util.stream.Stream;

public class PrioritizedConfigurationSource extends CompositeConfigurationSource {

    PrioritizedConfigurationSource(List<ConfigurationSource> sources) {

        super(sources);
    }

    PrioritizedConfigurationSource(SortedMap<ConfigurationSource, Namespace> sources) {
        super(sources);
    }

    @Override
    protected Stream<NamespacedWrapper> orderedSources(List<NamespacedWrapper> sources) {

        return sources.stream();
    }
}

File: lena-config/src/commonMain/kotlin/org/riversoforion/lena/config/PrioritizedConfigurationSource.kt
/*
 * Copyright (c) 2024-2026. Eric McIntyre / Rivers of Orion
 */
package org.riversoforion.lena.config

/**
 * A [CompositeConfigurationSource] that returns the first non-null value found across its
 * sources, in declaration order (first = highest priority).
 */
public class PrioritizedConfigurationSource internal constructor(
    sources: List<Pair<ConfigurationSource, Namespace?>>,
) : CompositeConfigurationSource(sources) {

    override fun orderedSources(
        all: List<Pair<ConfigurationSource, Namespace?>>,
    ): List<Pair<ConfigurationSource, Namespace?>> = all
}
